### ISSUE TITLE
_parse_connection_string() can't handle a shared_access_key with an '=' in it

### ISSUE THREAD

**Issue by rpack78** [ytnobody/p5-Net-Azure-Authorization-SAS#1]
Was attempting to use this module with a shared access key that had an '=' in it and was not able to authenticate with the service bus.  I noticed that there is a split on '=' that is breaking it.
`my %parsed = (map {split '=', $_, 2} split(';', $string));`

**Comment by ytnobody** [comment]
@rpack78 
Thank you for reporting.

I tried it with my service bus account. 

```
### try.pl
use strict;
use warnings;
use Data::Dumper;
use Net::Azure::Authorization::SAS;

my $str = 'Endpoint=sb://ytnobody-test-bus.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=4GCyRuLT8ERtQ+Pu2TvFwGADWXk1AD5sRlvY+vz53Gg=';
warn Dumper(+{Net::Azure::Authorization::SAS->_parse_connection_string($str)});
```

Then, result of `perl ./try.pl` is following.

```
$VAR1 = {
          'shared_access_key' => '4GCyRuLT8ERtQ+Pu2TvFwGADWXk1AD5sRlvY+vz53Gg=',
          'shared_access_key_name' => 'RootManageSharedAccessKey',
          'endpoint' => 'sb://ytnobody-test-bus.servicebus.windows.net/'
        };
```

It looks fine. But I feel strangeness about `a shared access key that had an '='`.

Typically, SharedAccessKey that contained in connection string terminates with '='. 
But, _parse_connection_string() works exactly when '=' is in middle of SharedAccessKey too.

```
### try2.pl
use strict;
use warnings;
use Data::Dumper;
use Net::Azure::Authorization::SAS;

my $str = 'Endpoint=sb://ytnobody-test-bus.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=4GCyRuLT8=ERt=Q+=P=u2TvFwGADWXk1AD5sRlvY+vz53Gg=';
warn Dumper(+{Net::Azure::Authorization::SAS->_parse_connection_string($str)});
```

Result of `perl ./try2.pl` is followings.

```
$VAR1 = {
          'shared_access_key_name' => 'RootManageSharedAccessKey',
          'shared_access_key' => '4GCyRuLT8=ERt=Q+=P=u2TvFwGADWXk1AD5sRlvY+vz53Gg=',
          'endpoint' => 'sb://ytnobody-test-bus.servicebus.windows.net/'
        };
```

Please tell me a connection string you used with partly masked if you can.

Best regards.

**Comment by rpack78** [comment]
I may have done something wrong, then.  This is the code I used that was giving me an invalid token:
```
#!/usr/bin/env perl
use Modern::Perl;
use Net::Azure::Authorization::SAS;

my $sas = Net::Azure::Authorization::SAS->new(
    connection_string => 'SharedAccessSignature=b5diJ7mrJEvYHz1ooEoEdPg9nXhmp7fiq4hajLHUoeE=',
    shared_access_key_name => 'send',
    shared_access_key => '',
);
my $token = $sas->token('https://removed-for-privacy.servicebus.windows.net/test');

say $token;
```
When I changed it to this it started working:
```
#!/usr/bin/env perl
use Modern::Perl;
use URI;
use URI::Escape 'uri_escape';
use Digest::SHA 'hmac_sha256';
use MIME::Base64 'encode_base64';

my $url = 'https://removed-for-privacy.servicebus.windows.net/test';
my $shared_access_key = 'b5diJ7mrJEvYHz1ooEoEdPg9nXhmp7fiq4hajLHUoeE=';
my $shared_access_key_name = 'send';

my $uri = URI->new($url);
my $target_uri = lc(uri_escape(lc(sprintf("%s://%s%s", $uri->scheme, $uri->host, $uri->path))));
my $expire_time = time + 3600;
my $to_sign = "$target_uri\n$expire_time";
my $signature = encode_base64(hmac_sha256($to_sign, $shared_access_key));
chomp $signature;
my $token = sprintf 'SharedAccessSignature sr=%s&sig=%s&se=%s&skn=%s', $target_uri, uri_escape($signature), $expire_time, $shared_access_key_name;

say $token;

**Comment by ytnobody** [comment]
@rpack78 

Maybe, your connection string is `Endpoint=https://removed-for-privacy.servicebus.windows.net/;SharedAccessKeyName=send;SharedAccessKey=b5diJ7mrJEvYHz1ooEoEdPg9nXhmp7fiq4hajLHUoeE=` .

```
#!/usr/bin/env perl
use Modern::Perl;
use Net::Azure::Authorization::SAS;

my $sas = Net::Azure::Authorization::SAS->new(
    connection_string => 'Endpoint=https://removed-for-privacy.servicebus.windows.net/;SharedAccessKeyName=send;SharedAccessKey=b5diJ7mrJEvYHz1ooEoEdPg9nXhmp7fiq4hajLHUoeE=',
);
my $token = $sas->token('https://removed-for-privacy.servicebus.windows.net/test');

say $token;
```

Another one.

```
#!/usr/bin/env perl
use Modern::Perl;
use Net::Azure::Authorization::SAS;

my $sas = Net::Azure::Authorization::SAS->new(
    connection_string      => '',
    shared_access_key      => 'b5diJ7mrJEvYHz1ooEoEdPg9nXhmp7fiq4hajLHUoeE=',
    endpoint               => 'https://removed-for-privacy.servicebus.windows.net/',
    shared_access_key_name => 'send',
);
my $token = $sas->token('https://removed-for-privacy.servicebus.windows.net/test');

say $token;
```

// To bypass `connection_string` is not cool. I think N::A::A::SAS should supports non connection_string style.

**Comment by rpack78** [comment]
Thank you, I will try this out today.

**Comment by ytnobody** [comment]
Today, I released 0.03.

https://twitter.com/cpan_new/status/994029971730415616

In new version, we can write as following.

```
#!/usr/bin/env perl
use Modern::Perl;
use Net::Azure::Authorization::SAS;

my $sas = Net::Azure::Authorization::SAS->new(
    endpoint               => 'https://removed-for-privacy.servicebus.windows.net/',
    shared_access_key      => 'b5diJ7mrJEvYHz1ooEoEdPg9nXhmp7fiq4hajLHUoeE=',
    shared_access_key_name => 'send',
);
my $token = $sas->token('https://removed-for-privacy.servicebus.windows.net/test');

say $token;
```

Happy Hacking.

**Comment by rpack78** [comment]
This is working great.  Thank you!
